tensorflow/python/ops/map_ops.py

```diff
@@ -66,10 +66,14 @@ def LookupGrad(op, dval):
 def InsertGrad(op, dmap):
   _, k, v = op.inputs
   key_grad = None
-  value_grad = control_flow_ops.cond(
-      tensor_map_has_key(dmap, k), lambda: tensor_map_lookup(dmap, k, v.dtype),
-      lambda: array_ops.zeros_like(v))
-  map_grad = control_flow_ops.cond(
-      tensor_map_has_key(dmap, k),
-      lambda: tensor_map_erase(dmap, k, v.dtype)[0], lambda: dmap)
+  (value_grad, map_grad) = control_flow_ops.cond(tensor_map_has_key(dmap, k),
+                                                 lambda: (tensor_map_lookup(dmap, k, v.dtype), tensor_map_erase(dmap, k, v.dtype)[0]),
+                                                 lambda: (array_ops.zeros_like(v), dmap))
   return map_grad, key_grad, value_grad
+
+@ops.RegisterGradient("TensorMapErase")
+def EraseGrad(op, dmap, dval):
+  _, k = op.inputs
+  key_grad = None
+  map_grad = tensor_map_insert(dmap, k, dval)
```

Comment:
Another alternative would be to not return the erased tensor from `TensorMapErase`. Then `EraseGrad` would simply return `dmap`.

@mdanatg what do you think?

Reply:
Yes, I think this makes a good reason to tilt the balance in favor of not returning the erased value.